cpp/src/io/parquet/page_data.cu

```diff
@@ -759,6 +759,24 @@ inline __device__ void gpuOutputInt64Timestamp(volatile page_state_s* s, int src
   *dst = ts;
 }
 
+/**
+ * @brief Output a byte array as int.
+ *
+ * @param[in] ptr Pointer to the byte array
+ * @param[in] len Byte array length
+ * @param[out] dst Pointer to row output data
+ */
+template <typename T>
+__device__ void gpuOutputByteArrayAsInt(char const* ptr, int32_t len, T* dst)
+{
+  T unscaled = 0;
+  for (auto i = 0; i < len; i++) {
+    uint8_t v = ptr[i];
+    unscaled  = (unscaled << 8) | v;
+  }
+  *dst = unscaled;
+}
+
```

Comment:
Right now, this function is only called with an `__int128_t *` `dst`. Does it have to be templated?

Relatedly, presumably there are some restrictions on valid parquet files? What if `len > 16` so that the left shift would overflow?

Reply:
Actually the parquet spec has no limit on the size of these arrays.

https://github.com/apache/parquet-format/blob/master/LogicalTypes.md

>    * fixed_len_byte_array: precision is limited by the array size. Length n can store <= floor(log_10(2^(8*n - 1) - 1)) base-10 digits
>   * binary: precision is not limited, but is required. The minimum number of bytes to store the unscaled value should be used.

We are imposing a 128-bit limit ourselves. That said I don't think it is common to have anything over that, so not making it templated should be fine.

Reply:
Good catch. I started with template since I tried narrow type in the first place. i.e. using the smallest decimal type based on its required precision. It didn't work out due to accuracy losses (still not sure why)